chips/rp2040/src/pwm.rs

```diff
@@ -0,0 +1,1354 @@
+/// is detected on pin B.
+pub enum DivMode {
+    FreeRunning,
+    High,
+    Rising,
+    Falling,
+}
+
+/// Channel identifier
+///
+/// There are a total of 8 eight PWM channels.
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum ChannelNumber {
+    Ch0,
+    Ch1,
+    Ch2,
+    Ch3,
+    Ch4,
+    Ch5,
+    Ch6,
+    Ch7,
+}
+
+const CHANNEL_NUMBERS: [ChannelNumber; NUMBER_CHANNELS] = [
+    ChannelNumber::Ch0,
+    ChannelNumber::Ch1,
+    ChannelNumber::Ch2,
+    ChannelNumber::Ch3,
+    ChannelNumber::Ch4,
+    ChannelNumber::Ch5,
+    ChannelNumber::Ch6,
+    ChannelNumber::Ch7,
+];
+
+/// Each GPIO pin can be configured as a PWM pin.
+/// The following table shows the mapping between GPIO pins and PWM pins:
+///
+/// | GPIO  | PWM |
+/// | ----- | --- |
+/// | 0     | 0A  |
+/// | 1     | 0B  |
+/// | 2     | 1A  |
+/// | 3     | 1B  |
+/// | 4     | 2A  |
+/// | 5     | 2B  |
+/// | 6     | 3A  |
+/// | 7     | 3B  |
+/// | 8     | 4A  |
+/// | 9     | 4B  |
+/// | 10    | 5A  |
+/// | 11    | 5B  |
+/// | 12    | 6A  |
+/// | 13    | 6B  |
+/// | 14    | 7A  |
+/// | 15    | 7B  |
+/// | 16    | 0A  |
+/// | 17    | 0B  |
+/// | 18    | 1A  |
+/// | 19    | 1B  |
+/// | 20    | 2A  |
+/// | 21    | 2B  |
+/// | 22    | 3A  |
+/// | 23    | 3B  |
+/// | 24    | 4A  |
+/// | 25    | 4B  |
+/// | 26    | 5A  |
+/// | 27    | 5B  |
+/// | 28    | 6A  |
+/// | 29    | 6B  |
+///
+/// **Note**:
+///
+/// + The same PWM output can be selected on two GPIO pins. The same signal will appear on each
+/// GPIO.
+/// + If a PWM B pin is used as an input, and is selected on multiple GPIO pins, then the PWM
+/// channel will see the logical OR of those two GPIO inputs
+impl From<RPGpio> for ChannelNumber {
+    fn from(gpio: RPGpio) -> Self {
+        match gpio as u8 >> 1 & 7 {
+            0 => ChannelNumber::Ch0,
+            1 => ChannelNumber::Ch1,
+            2 => ChannelNumber::Ch2,
+            3 => ChannelNumber::Ch3,
+            4 => ChannelNumber::Ch4,
+            5 => ChannelNumber::Ch5,
+            6 => ChannelNumber::Ch6,
+            7 => ChannelNumber::Ch7,
+            // This branch can't be reached due to logical AND
+            _ => panic!("Unreachable branch"),
+        }
+    }
+}
+
+/// Identifier for a channel pin
+///
+/// Each PWM channel has two pins: A and B.  
+/// Pin A is always configured as an output pin.  
+/// Pin B is configured as an output pin when running in free running mode. Otherwise, it is
+/// configured as an input pin.
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum ChannelPin {
+    A,
+    B,
+}
+
+/// Check ChannelNumber implementation for more details
+impl From<RPGpio> for ChannelPin {
+    fn from(gpio: RPGpio) -> Self {
+        match gpio as u8 & 1 {
+            0 => ChannelPin::A,
+            1 => ChannelPin::B,
+            // This branch can't be reached due to logical AND
+            _ => panic!("Unreachable branch"),
```

Comment:
I suggest removing the panic this adds to code size.

```suggestion
            _ => {},
```

Reply:
I cannot add your suggestion, since the return type of **{}** is **()**. However, I changed the match instruction by replacing the last valid arm with a catch-all arm. If you don't like the change, I can replace the match instruction with an if else instruction.

Reply:
Ah, got it. Looks good with default.